src/shogun/base/subset_iterators.h

```diff
@@ -0,0 +1,128 @@
+            (std::is_base_of<T, CFeatures>::value) || (std::is_base_of<T, CLabels>::value)> * = nullptr>
+    class SubsetIterator {
+    public:
+        SubsetIterator(T *pointer) {
+            if (pointer->get_subset_stack()->get_last_subset() != nullptr)
+                m_subset = pointer->get_subset_stack()->get_last_subset()->get_subset_idx();
+            m_idx = 0;
+            m_argsorted_subset = CMath::argsort(m_subset);
+            m_argsorted_subset_iter = m_argsorted_subset.begin();
+            m_ptr = pointer;
+            SG_REF(m_ptr)
+        }
+
+        ~SubsetIterator() {
+            SG_UNREF(m_ptr)
+        }
+
+        SubsetIterator operator++(int) {
+            m_argsorted_subset_iter++;
+            m_idx++;
+            return *this;
+        }
+
+        SubsetIterator& operator++() {
+            m_argsorted_subset_iter++;
+            m_idx++;
+            return *this;
+        }
+
+        const auto operator*() {
+            index_t idx;
+            if (m_ptr->get_subset_stack()->get_last_subset() != nullptr)
+                idx = m_subset[*m_argsorted_subset_iter];
+            else
+                idx = m_idx;
+            if constexpr (std::is_base_of<CFeatures, T>::value)
+                return m_ptr->get_feature_vector(idx);
+            if constexpr (std::is_base_of<CLabels, T>::value)
+                return m_ptr->get_label(idx);
+        }
+
+	template <
+	    typename T, std::enable_if<
+	                    (std::is_base_of<T, CFeatures>::value) ||
+	                    (std::is_base_of<T, CLabels>::value)>* = nullptr>
+	class SubsetIterator
+	{
+	public:
+		using iterator_category = std::forward_iterator_tag;
+		using value_type = typename return_type_subset_iterator<T>::type;
+        using difference_type = index_t;
+        // not a reference but is used by stl algorithms to check type
+		using reference = typename return_type_subset_iterator<T>::type;
+		using pointer = typename return_type_subset_iterator<T>::type*;
+
+		SubsetIterator(T* pointer, index_t idx = 0)
+		{
+			if (pointer->get_subset_stack()->get_last_subset() != nullptr)
+				m_subset = pointer->get_subset_stack()
+				               ->get_last_subset()
+				               ->get_subset_idx();
+			m_idx = idx;
+			// need to keep this variable so that the vector destructor isn't
+			// called
+			m_argsorted_subset = CMath::argsort(m_subset);
+			m_argsorted_subset_iter = m_argsorted_subset.begin() + idx;
+			m_ptr = pointer;
+		}
+
+		SubsetIterator& operator++()
+		{
+			m_argsorted_subset_iter++;
+			m_idx++;
+			return *this;
+		}
+
+		SubsetIterator operator++(int)
+		{
+			SubsetIterator retval(*this);
+            ++(*this);
+			return retval;
+		}
+
+		bool operator==(const SubsetIterator<T>& other)
+		{
+			return m_ptr == other.m_ptr && m_idx == other.m_idx;
+		}
+
+		bool operator!=(const SubsetIterator<T>& other)
+		{
+			return !(*this == other);
+		}
+
+		const reference operator*()
```

Comment:
from what i understand this is legal c++ (at least for now and the compilers we support). dereferencing an iterator doesn't have to return a `const obj&`.. but this can cause issues (see http://ericniebler.com/2015/01/28/to-be-or-not-to-be-an-iterator/)

Reply:
but then this won't compile: `const auto& val = *it;`

Reply:
lol, all new stuff  for me :)